### ISSUE TITLE
Overlapping pom dependency management entries can cause a wrong dependency scope in Gradle 4.5

### ISSUE THREAD

**Issue by wreulicke** [gradle/gradle#4202]
### Expected Behavior

I added dependency of "org.zalando:logbook-spring-boot-starter:1.5.3" into my module.
slf4j-simple and lombok is not resloved as dependencies of "logbook-spring-boot-starter" .

I seems to have no that dependency in [the module](https://search.maven.org/#artifactdetails%7Corg.zalando%7Clogbook-spring-boot-starter%7C1.5.3%7Cjar)

### Current Behavior

In gradle 4.5, just resolved  as dependencies.
But in gradle 4.4, not resolved.

### Steps to Reproduce (for bugs)

Next gradle task shows change of dependencies between gradle 4.4 and 4.5.

```
gradle dependencies
```

### Your Environment

 * Build scan URL: private repository. (try to reproduce other public repository)

**Comment by wreulicke** [comment]
Related? https://github.com/gradle/gradle/issues/4171

**Comment by jjohannes** [comment]
@wreulicke thanks for reporting. I can reproduce this. It is a regression.

Looks like `json-path-assert:2.2.0` is somehow promoted to the `compile` scope although it is defined in `test` scope in the [parent POM](http://central.maven.org/maven2/org/zalando/logbook-parent/1.5.3/logbook-parent-1.5.3.pom).

**Comment by jjohannes** [comment]
Actually the pom `org.zalando.logbook-parent:logbook-parent` linked above imports `org.springframework.boot:spring-boot-dependencies`. Which also has an entry for `json-path-assert` - without scope.

We treat entries with different scopes now separately. Before, one entry would just override the other. Which could lead to information loss, if we have two entries with different scopes. See: https://github.com/gradle/gradle/commit/a82cf4ac0b40d25cb12218df669be14259f9a675

I think the issue now is that we treat the entry without scope as "scope == compile", independent of whether we saw other entries or not. But we should only do that if no scope is defined at all in any entry.

**Comment by jjohannes** [comment]
After investigating further, I think the change (https://github.com/gradle/gradle/commit/a82cf4ac0b40d25cb12218df669be14259f9a675) was a mistake. Even if we fix the case here, there are other cases where we would need to "merge" information from multiple dependency management entries to have the "correct" (i.e. mavens) behaviour. But that would mean that the behaviour changes for some cases.

If we do this more involved change (I'll open a separate issue for it), then we should do it for 5.0.

For now, I will revert https://github.com/gradle/gradle/commit/a82cf4ac0b40d25cb12218df669be14259f9a675 for 4.6.

@wreulicke is this actually blocking you from using 4.5 or is it something you "just noticed"? Can you work around it with an exclude?

**Comment by wreulicke** [comment]
@jjohannes thanks for your comments.

> Even if we fix the case here, there are other cases where we would need to "merge" information from multiple dependency management entries to have the "correct" (i.e. mavens) behaviour. But that would mean that the behaviour changes for some cases.

I see. but, I think it is wrong to add `real` dependency, when module add to dependency management in `pom.xml`.
my understand is correct?

> @wreulicke is this actually blocking you from using 4.5 or is it something you "just noticed"? Can you work around it with an exclude?

blocking. But I can work around for few part.

In my case, my runtime dependencies was contaminated due to this issue.
So, this contamination leads to break a application because there are two slf4j implementation in my classpath (ex. logback and slf4j-simple).

In above case, I can work around. Following.

```
configurations.all {
  exclude module: 'slf4j-simple'
}
```

My runtime dependencies was contaminated by other dependency `lombok`, yet.
lombok may be harmless in runtime.
But there are no solution to this contaimination...

Do you have any solution?

**Comment by wreulicke** [comment]
> I see. but, I think it is wrong to add real dependency, when module add to dependency management in pom.xml.

I misunderstood.
This contamination is occured when there are some dependency in `<dependencyManagement>` and  dependency with other scope  in `<dependencies>`?

**Comment by jjohannes** [comment]
To clarify: The issue can only appears if one of the `pom`s defines a dependency without scope and where a scope is added via `<dependencyManagement/>`. And where another  `<dependencyManagement/>` entry exists with a different scope. 

The revert of the change above will bring us back to what we had prior to 4.5: 
We always pick the "closest" dependency management entry - which is the one with test scope in your setting - and ignore all the other entries.

But it looks like it's worse for you than I thought. There are multiple test dependencies following this pattern - including `lombock` as you noticed.
You can work around this by using one of our experimental new APIs for _component metadata rules_. These allow you to adjust the metadata and remove the dependencies. Here is a script that should remove all the wrong entries:

```
dependencies {
    components {
        all { metadata ->
            workaroundGradle4202(metadata, "com.jayway.jsonpath", "")
            workaroundGradle4202(metadata, "org.projectlombok", "")
            workaroundGradle4202(metadata, "org.mockito", "")
            workaroundGradle4202(metadata, "org.springframework", "spring-test")
            workaroundGradle4202(metadata, "org.slf4j", "jcl-over-slf4j")
            workaroundGradle4202(metadata, "org.slf4j", "slf4j-nop")
            workaroundGradle4202(metadata, "org.slf4j", "slf4j-simple")
        }
    }
}

def workaroundGradle4202(ComponentMetadataDetails metadata, String testDependencyGroup, String testDependencyName) {
    metadata.withVariant("default") {
        withDependencies { it.removeAll {
            it.group == testDependencyGroup && (testDependencyName.empty || it.name == testDependencyName)
        }}
    }
    metadata.withVariant("runtime") {
        withDependencies { it.removeAll {
            it.group == testDependencyGroup && (testDependencyName.empty || it.name == testDependencyName)
        }}
    }
}
```
_Note that this API is not yet finished and officially released and will change in future versions. But this way it works for 4.5 as workaround._

Excluding all the dependencies as you started to do probably also work if you exclude **all** the ones listed above.

**Comment by jjohannes** [comment]
We have a `4.5.1` release scheduled for Monday that will include a fix for this.

PR with fix (revert): #4244

**Comment by wreulicke** [comment]
> We have a 4.5.1 release scheduled for Monday that will include a fix for this.

@jjohannes thanks to your quickly help. 👍
